Log Overhaul

cosmostevens opened the pull request:
Changes:
- Filter Tab Pane
     - Filters are now displayed in a tab pane on the left
     - Tab headers darken if filters are applied
     - Log name filter and log level filters were moved into filter panes
- LogDetailsPane:
     - Selected logs are now collapsible
     - Logs can be individually or collectively popped out into a separate frame
     - Font/style is consistent for all details
- LogNameList: 
     - Inherits from FilterList 
     - Lists now sortable by full logger name or by subname.
- FilterList:
     - Move from threads to utils
- models.kt
     - Added LogFilterPanel interface (implemented by LoggerMDCs, LoggerTimesPanel, LogLevelsList, and LogNameList)
- Wrapper log filenames of log(.n) will now work up to n == 20
- Changed Wrapper log encoding to ISO_8859_1
- Replaced minimum logging level with a filter list of all logging levels
- Changed SystemLogsEvent to SystemLogEvent
- Added theme change listeners
- Removed redundant filter headers
- Resolved package name conflicts

Additions:
- Marking:
     - Added ability to mark logs
     - Added ability to only show marked logs 
- Context Menu filter actions:
     - Added ability to mark or filter all by: Logger Name, Level, Thread (if it is a system log), Message, and Stacktrace (complete or partial) 
     - Filters or 'mark all' events can be removed if they have been applied
     - JMenu items will show checks if filter is already applied or all log events with a specific value is already marked, otherwise they will remain unchecked
- LoggerMDCs:
     - Added ability to add/remove MDC filters if the log is a system log from the MDC filter panel
     - Filters can be inverted or normal
     - Filters can be enabled/disabled
     - Increased button font size
     - Added smart selection when rows are removed
     - Added context menu with removal option
- LoggerTimesPanel:
     - Added ability to filter logs by date and time down to the millisecond
     - Changes are applied automatically
     - Can be enabled/disabled
     - Can be reset to default
     - Replaced JComboBoxes with spinners
     - Spinners have value dragging ability
- LogLevelsList:
     - a FilterList and LogFilterPanel that stores log level filter options

paul-griffith commented on src/main/kotlin/io/github/inductiveautomation/kindling/log/LogEvent.kt:
```diff
@@ -3,31 +3,37 @@ package io.github.inductiveautomation.kindling.log
 
 sealed interface LogEvent {
+    var marked: Boolean
     val timestamp: Instant
```
Can put `marked` at the end with a default value of false so you don't have to change all the constructor calls

paul-griffith commented on src/main/kotlin/io/github/inductiveautomation/kindling/log/LogNameList.kt:
```diff
@@ -0,0 +1,120 @@
+    var isShowFullLoggerNames = false
+        set(value) {
+            field = value
+            updateCellRenderer()
```
Is resetting the renderer entirely actually necessary here?

paul-griffith commented on src/main/kotlin/io/github/inductiveautomation/kindling/log/models.kt:
```diff
@@ -71,52 +95,69 @@ class SystemLogsColumns(panel: LogPanel) : ColumnList<SystemLogsEvent>() {
+        add(Timestamp)
+    }
+
+    abstract val filterableColumns: List<Column<T, out Any?>>
```
`out Any?` can be written as just `*`

paul-griffith commented on src/main/kotlin/io/github/inductiveautomation/kindling/log/models.kt:
```diff
@@ -71,52 +95,69 @@ class SystemLogsColumns(panel: LogPanel) : ColumnList<SystemLogsEvent>() {
+        add(Thread)
+    }
+
+    override val filterableColumns = listOf(
```
Both subclasses have the exact same values in these lists - can be combined into one?

paul-griffith commented on src/main/kotlin/io/github/inductiveautomation/kindling/log/LogPanel.kt:
```diff
@@ -120,13 +418,37 @@ class LogPanel(
+
+    private fun JTabbedPane.updateBackgroundAt(condition: Boolean, index: Int) {
+        if (condition) {
+            setBackgroundAt(index, background.darker())
```
background.darker() barely shows up on MacOS dark theme - any reason to go with that over a more 'prominent' color from the LaF?

cosmostevens replied:
I just wanted to add something simple that would work across all of the themes. Honestly I didn't test on MacOS though, I can find something a from the LaF though.

paul-griffith commented on src/main/kotlin/io/github/inductiveautomation/kindling/log/LogNameList.kt:
```diff
@@ -0,0 +1,120 @@
+
+    private fun updateCellRenderer() {
+        cellRenderer = listCellRenderer<Any?> { _, value, _, _, _ ->
+            text = when (value) {
```
No more tooltips on these entries?

cosmostevens replied:
Oops, it looks like they got removed in the refactor. I can throw them back in.

paul-griffith commented on src/main/kotlin/io/github/inductiveautomation/kindling/log/WrapperLogView.kt:
```diff
@@ -47,15 +47,19 @@ class WrapperLogView(
         require(paths.isNotEmpty()) { "Must provide at least one path" }
         val events = paths.flatMap { path ->
-            path.useLines { lines -> LogPanel.parseLogs(lines) }
+            path.useLines(Charsets.ISO_8859_1) { lines -> LogPanel.parseLogs(lines) }
```
I'm somewhat leery of this change... it'd be nice to be able to select a charset in the file picker, perhaps, as a 'decoration' (supported in JFileChooser)

Joshlha replied:
Being able to select the charset is something we discussed. This is sort of a band aid which works with all of the troublesome wrappers we came across and didn't break anything. 

Based on our testing, this is definitely working. But ideally a charset picker I think is best. We could probably catch the exception and prompt the user to select a different charset

paul-griffith replied:
There won't be an exception if the charset you assume is wrong, is the thing. Any international user (e.g. a wrapper log from a Japanese system) will load just fine...but some/all of the text will be totally mangled. See [Mojibake](https://en.wikipedia.org/wiki/Mojibake).

paul-griffith replied:
Charset could now live in preferences, with the default being ISO-8859-1

paul-griffith commented on src/main/kotlin/io/github/inductiveautomation/kindling/log/WrapperLogView.kt:
```diff
@@ -47,15 +47,19 @@ class WrapperLogView(
 }
 
 object LogViewer : MultiTool, ClipboardTool {
+    private const val MAX_EXTENSION_INDEX = 20
```
Has this actually come up before? It's not like counts > 5 can't be opened, they just have to switch to 'all files', right?

cosmostevens replied:
Surprisingly, yes. It's pretty common to get up to 10. I asked around and someone got 100 at one point (which is almost absurd). The all files method does work, but this makes it easier for drag and drop.

paul-griffith reviewed (changes requested):
I could probably keep nitpicking, but these are my biggest concerns. Love the work overall, though!

Joshlha commented on src/main/kotlin/io/github/inductiveautomation/kindling/log/LogDetailsPane.kt:
```diff
@@ -0,0 +1,245 @@
+                }
+                pre { 
+                    font-size: 10px; 
+                    font-family: Segoe
```
I think the justification was "Dialog is ugly" but we can probably just use it

cosmostevens replied:
I believe I originally made this switch because the log details had two different fonts, one for the stacktrace and one for the message. This was an attempt to make it consistent, however this issue is accounted for in other ways now that the whole details pane is restructured.

paul-griffith replied:
I'm open to including a FOSS font like one of the ones from FlatLaf if that's preferred:
https://github.com/JFormDesigner/FlatLaf/tree/main/flatlaf-fonts

cosmostevens commented:
- Added FlatLaf Roboto Dependency and changed universal font to Roboto
- Changed event.mdc from Map<String, String> to List<MDC>
- Passed keyList into the JComboBox constructor to avoid repeatedly resizing the internal Vector in the DefaultComboboxModel
- Created keyValueMap : Map<String, List<MDC>> to avoid expensive operations when the MDC keyMenu selection is changed
- Condensed filterableColumns and markableColumns into filterableColumns to avoid redundant lists
- Changed the background color for filter tabs whose filter is applied to TabbedPane.focusColor
- Set isOneTouchExpandable = true for the filters so that they can be easily minimized/expanded
- Added the accidentally removed tooltips for loggerNames
- Added a Charset Encoding selector to the Open JFileChooser with default set to ISO_8859_1; Drag and drop will use the last selected Encoding
